akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedJournalInteractions.scala

```diff
@@ -73,16 +70,41 @@ private[akka] trait EventsourcedJournalInteractions[C, E, S] {
 
   protected def replayEvents(fromSeqNr: Long, toSeqNr: Long): Unit = {
     setup.log.debug("Replaying messages: from: {}, to: {}", fromSeqNr, toSeqNr)
-    // reply is sent to `selfUntypedAdapted`, it is important to target that one
     setup.journal ! ReplayMessages(fromSeqNr, toSeqNr, setup.recovery.replayMax, setup.persistenceId, setup.selfUntyped)
   }
 
-  protected def returnRecoveryPermit(setup: EventsourcedSetup[_, _, _], reason: String): Unit = {
-    setup.log.debug("Returning recovery permit, reason: " + reason)
-    // IMPORTANT to use selfUntyped, and not an adapter, since recovery permitter watches/unwatches those refs (and adapters are new refs)
-    setup.persistence.recoveryPermitter.tell(RecoveryPermitter.ReturnRecoveryPermit, setup.selfUntyped)
+  protected def requestRecoveryPermit(): Unit = {
+    setup.persistence.recoveryPermitter.tell(RecoveryPermitter.RequestRecoveryPermit, setup.selfUntyped)
+  }
+
+  /** Intended to be used in .onSignal(returnPermitOnStop) by behaviors */
+  protected def returnPermitOnStop: PartialFunction[(ActorContext[InternalProtocol], Signal), Behavior[InternalProtocol]] = {
+    case (_, PostStop) ⇒
+      tryReturnRecoveryPermit("PostStop")
+      Behaviors.stopped
+    case (_, PreRestart) ⇒
+      // TODO was not entirely sure if it's needed here as well
+      tryReturnRecoveryPermit("PostStop")
+      Behaviors.stopped
   }
 
+  /** Mutates setup, by setting the `holdingRecoveryPermit` to false */
+  protected def tryReturnRecoveryPermit(reason: String): Unit = {
```

Comment:
this is called from `onRecoveryFailure`, which shouldn't be necessary since it's done in PreRestart and PostStop.

It must still be done from `onRecoveryCompleted` ofc.

Reply:
Yes though there it can include beffer information about the reason: ` tryReturnRecoveryPermit("on replay failure: " + cause.getMessage)` o to keep? Due to the flag it's not incorrect, it will be a no-op in PostStop

Reply:
ok